torch/_dynamo/variables/nn_module.py

```diff
@@ -303,33 +303,29 @@ def call_method(
         key = self.module_key
         module = tx.output.get_submodule(key)
 
-        if name == "__call__":
-            # TODO(whc)  do we really need this special case?
+        if name == "_call_impl":
+            # Example: `self.layer.__call__(x)`
+            # This is used for explicit calling `__call__` in a forward function.
+            # Dynamo inlines `__call__`, includes hooks.
             return self.call_function(tx, args, kwargs)
-        elif name == "forward":
-            # TODO(whc)
-            # This is the old special case moved to a new place.  (copy from call_function below)
-            # Old behavior: we'd route "forward" meth call to 'call_function', which inlined forward.
-            # New behavior: since call_function now hits '__call__', forward would fall through to 'wrap_proxy' below,
-            # instead of being inlined.  What should we do about this?
-            #   1) all methods get inlined now at the bottom of this call_method, instead of put into the graph as calls
-            #   2) we maintain this special case just for forward
-            assert self.source, (
-                "Must provide a valid source in order to inline, "
-                "since inlined function may have default args which must be guarded."
-            )
-            fn = module.forward.__func__
-            assert istype(fn, types.FunctionType)
-            options["source"] = AttrSource(
-                AttrSource(self.source, "forward"), "__func__"
-            )
-            args = [self] + args
+        elif name == "forward" and is_allowed(module.__class__):
```

Comment:
are you deleting the handling for case `name=="forward" and not allowed`?

Reply:
Yes, see my comment there: ```For not allowed module, it has been inlined before getting here.```
Only the allowed module inlining will route here, because of:
https://github.com/pytorch/pytorch/blob/cdef4f073c751188b7737845e136b6cb65f1b4d8/torch/_dynamo/variables/functions.py#L287-L296

Reply:
hmm, you sure it covers all cases? Looks like it only does that under certain cases.  (not sure if module_attr is set on user-defined modules or what the meaning of that is)

```
   module_attr is not None
   and module_attr.startswith("torch.nn.")
   or self.is_constant
```

Also, it's a little bit fragile to assume that the only way you can get into this codepath is by the other codepath you pointed.  What if someone added a new codepath that calls this in a seeming valid way, but doesn't know about this assumption?

Reply:
I think the reason of routing this case to ```self.obj.call_method``` is: we don't want to do general inlining for the function calls, and we believe inside of ```self.obj.call_method``` have better handling than basic inlining. I just don't want to have too much branches that making the code hard to maintain. I think ```module_attr.startswith("torch.nn.") ``` means allowed module?

Reply:
I think by design, we should know for the common cases where we should route to. We should avoid adding too many different branches for the cases can be consolidated. That's my general idea, but I'm open to hear you thought.